### PR TITLE
Manage RBAC for openshift-controller-manager controllers

### PR DESCRIPTION
**Pull request by adambkaplan**
Update openshift-controller-manager-operator to directly manage the RBAC permissions for the openshift-controller-manager controllers. These are currently managed by legacy openshift-apiserver bootstrap code, which will be removed by https://github.com/openshift/openshift-apiserver/pull/209.

The RBAC for the following controllers will now be managed by the operator:

- serviceaccount-controller
- build-controller
- build-config-change-controller
- deployer-controller
- deploymentconfig-controller
- template-instance-controller
- template-instance-finalizer-controller
- origin-namespace-controller
- serviceaccount-pull-secrets-controller
- image-trigger-controller
- image-import-controller
- unidling-controller
- service-ingress-ip-controller
- default-rolebindings-controller

All new role names and role bindings will use the `system:openshift:openshift-controller-manager:*` naming scheme, to ensure that these roles are seamlessly applied to the cluster on upgrade.

### CODE REVIEW

File: pkg/operator/sync_openshiftcontrollermanager_v311_00.go [hunk 1]
```diff
@@ -38,41 +37,9 @@ func syncOpenShiftControllerManager_v311_00_to_latest(c OpenShiftControllerManag
...
-
-	for _, currResult := range directResourceResults {
-		if currResult.Error != nil {
-			errors = append(errors, fmt.Errorf("%q (%T): %v", currResult.File, currResult.Type, currResult.Error))
-			continue
-		}
 
-		if currResult.Changed && resourcesThatForceRedeployment.Has(currResult.File) {
-			forceRollout = true
-		}
```
Comment (adambkaplan) [inline pkg/operator/sync_openshiftcontrollermanager_v311_00.go:74]: @deads2k this was in the original implementation, do we need to check for changes to ocm's ServiceAccount object to force a rollout of the DaemonSet?

File: pkg/operator/starter.go [hunk 2]
```diff
@@ -105,10 +108,40 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
 		controllerConfig.EventRecorder,
 	)
 
+	staticResourceController := staticresourcecontroller.NewStaticResourceController(
```
Comment (deads2k) [inline pkg/operator/starter.go:111]: future me is strongly in favor a `directory/...` mechanism.

File: pkg/operator/sync_openshiftcontrollermanager_v311_00.go (hunk 1 above)
Comment (deads2k) [reply pkg/operator/sync_openshiftcontrollermanager_v311_00.go:74]: > @deads2k this was in the original implementation, do we need to check for changes to ocm's ServiceAccount object to force a rollout of the DaemonSet?

I cannot think of a good reason for me doing that.  It is possible that I was concerned about a delete and recreate of the serviceaccount causing new SA tokens to be produced.  I don't think I would worry about it anymore.

**Comment by deads2k** [comment]
This looks like a change in the correct direction.  Does the openshift-apiserver still have duplicates of these roles somewhere?

**Comment by adambkaplan** [comment]
> This looks like a change in the correct direction. Does the openshift-apiserver still have duplicates of these roles somewhere?

Yes - these roles are still in the bootstrap logic. I only think ingress-to-route controller was moved over when network-edge had to update their RBAC. See https://github.com/openshift/openshift-apiserver/blob/master/pkg/bootstrappolicy/controller_policy.go#L91

File: pkg/operator/starter.go (hunk 2 above)
Comment (adambkaplan) [reply pkg/operator/starter.go:111]: yes - there will be a _lot_ of controller RBAC moving over in this refactor.

File: bindata/v3.11.0/openshift-controller-manager/build-controller-clusterrole.yaml [hunk 3]
```diff
@@ -0,0 +1,109 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:build-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/build-controller-clusterrole.yaml:4]: @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/build-controller-clusterrolebinding.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:build-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/build-controller-clusterrolebinding.yaml:4]: @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrole.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:default-rolebindings-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrole.yaml:4]: @sttts @deads2k

File: bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrolebinding-deployer.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:deployer
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrolebinding-deployer.yaml:4]: @soltysh

File: bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrolebinding-image-builder.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:image-builder
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrolebinding-image-builder.yaml:4]: @openshift/openshift-team-build-api @sttts @deads2k

File: bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrolebinding-image-puller.yaml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:image-puller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrolebinding-image-puller.yaml:4]: @openshift/openshift-team-build-api @sttts @deads2k

File: bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrolebinding.yaml [hunk 9]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:default-rolebindings-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/default-rolebindings-controller-clusterrolebinding.yaml:4]: @sttts @deads2k

File: bindata/v3.11.0/openshift-controller-manager/deployer-controller-clusterrole.yaml [hunk 10]
```diff
@@ -0,0 +1,47 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:deployer-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/deployer-controller-clusterrole.yaml:4]: @soltysh

File: bindata/v3.11.0/openshift-controller-manager/deploymentconfig-controller-clusterrolebinding.yaml [hunk 11]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:deploymentconfig-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/deploymentconfig-controller-clusterrolebinding.yaml:4]: @soltysh

File: bindata/v3.11.0/openshift-controller-manager/image-import-controller-clusterrole.yaml [hunk 12]
```diff
@@ -0,0 +1,44 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:image-import-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/image-import-controller-clusterrole.yaml:4]: @dmage

File: bindata/v3.11.0/openshift-controller-manager/image-trigger-controller-clusterrole.yaml [hunk 13]
```diff
@@ -0,0 +1,76 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:image-trigger-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/image-trigger-controller-clusterrole.yaml:4]: @dmage @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/image-trigger-controller-clusterrolebinding.yaml [hunk 14]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:image-trigger-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/image-trigger-controller-clusterrolebinding.yaml:4]: @dmage @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/origin-namespace-controller-clusterrole.yaml [hunk 15]
```diff
@@ -0,0 +1,28 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:origin-namespace-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/origin-namespace-controller-clusterrole.yaml:4]: @sttts @deads2k

File: bindata/v3.11.0/openshift-controller-manager/origin-namespace-controller-clusterrolebinding.yaml [hunk 16]
```diff
@@ -0,0 +1,15 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  annotations:
+    rbac.authorization.kubernetes.io/autoupdate: "true"
+  creationTimestamp: null
+  name: system:openshift:openshift-controller-manager:origin-namespace-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/origin-namespace-controller-clusterrolebinding.yaml:7]: @sttts @deads2k

File: bindata/v3.11.0/openshift-controller-manager/service-ingress-ip-controller-clusterrole.yaml [hunk 17]
```diff
@@ -0,0 +1,27 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:service-ingress-ip-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/service-ingress-ip-controller-clusterrole.yaml:4]: @Miciah

File: bindata/v3.11.0/openshift-controller-manager/service-ingress-ip-controller-clusterrolebinding.yaml [hunk 18]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:service-ingress-ip-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/service-ingress-ip-controller-clusterrolebinding.yaml:4]: @Miciah

File: bindata/v3.11.0/openshift-controller-manager/serviceaccount-controller-clusterrole.yaml [hunk 19]
```diff
@@ -0,0 +1,31 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:serviceaccount-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/serviceaccount-controller-clusterrole.yaml:4]: @dmage @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/serviceaccount-controller-clusterrolebinding.yaml [hunk 20]
```diff
@@ -0,0 +1,11 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:serviceaccount-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/serviceaccount-controller-clusterrolebinding.yaml:4]: @dmage @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/serviceaccount-pull-secrets-controller-clusterrole.yaml [hunk 21]
```diff
@@ -0,0 +1,44 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:serviceaccount-pull-secrets-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/serviceaccount-pull-secrets-controller-clusterrole.yaml:4]: @dmage @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/serviceaccount-pull-secrets-controller-clusterrolebinding.yaml [hunk 22]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:serviceaccount-pull-secrets-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/serviceaccount-pull-secrets-controller-clusterrolebinding.yaml:4]: @dmage @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/template-instance-controller-clusterrole.yaml [hunk 23]
```diff
@@ -0,0 +1,17 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:template-instance-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/template-instance-controller-clusterrole.yaml:4]: @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/template-instance-controller-clusterrolebinding-admin.yaml [hunk 24]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:template-instance-controller:admin
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/template-instance-controller-clusterrolebinding-admin.yaml:4]: @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/template-instance-controller-clusterrolebinding.yaml [hunk 25]
```diff
@@ -0,0 +1,8 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:template-instance-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/template-instance-controller-clusterrolebinding.yaml:4]: @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/template-instance-finalizer-controller-clusterrole.yaml [hunk 26]
```diff
@@ -0,0 +1,11 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:template-instance-finalizer-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/template-instance-finalizer-controller-clusterrole.yaml:4]: @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/template-instance-finalizer-controller-clusterrolebinding-admin.yaml [hunk 27]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:template-instance-finalizer-controller:admin
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/template-instance-finalizer-controller-clusterrolebinding-admin.yaml:4]: @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/template-instance-finalizer-controller-clusterrolebinding.yaml [hunk 28]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:template-instance-finalizer-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/template-instance-finalizer-controller-clusterrolebinding.yaml:4]: @openshift/openshift-team-build-api

File: bindata/v3.11.0/openshift-controller-manager/unidling-controller-clusterrole.yaml [hunk 29]
```diff
@@ -0,0 +1,63 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:openshift:openshift-controller-manager:unidling-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/unidling-controller-clusterrole.yaml:4]: @sttts @deads2k @soltysh 

To be honest I don't know which team _should_ own this.

File: bindata/v3.11.0/openshift-controller-manager/unidling-controller-clusterrolebinding.yaml [hunk 30]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:openshift:openshift-controller-manager:unidling-controller
```
Comment (adambkaplan) [inline bindata/v3.11.0/openshift-controller-manager/unidling-controller-clusterrolebinding.yaml:4]: @sttts @deads2k @soltysh 

Ditto - unclear which team _should_ own this.

**Review by adambkaplan — COMMENTED** [review]
/hold

Need review from relevant stakeholders before we merge this.

File: bindata/v3.11.0/openshift-controller-manager/image-import-controller-clusterrole.yaml [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+rules:
+- apiGroups:
+  - ""
```
Comment (dmage) [inline bindata/v3.11.0/openshift-controller-manager/image-import-controller-clusterrole.yaml:7]: afaik you don't need the apiGroup `""` for imagestreams

Comment (adambkaplan) [reply bindata/v3.11.0/openshift-controller-manager/image-import-controller-clusterrole.yaml:7]: That's probably a legacy of using non-groupified APIs from way back. @deads2k is it safe for us to remove?

Comment (deads2k) [reply bindata/v3.11.0/openshift-controller-manager/image-import-controller-clusterrole.yaml:7]: > That's probably a legacy of using non-groupified APIs from way back. @deads2k is it safe for us to remove?

I suggest keeping it for the purposes of this PR.

**Comment by deads2k** [comment]
If you have kept the same names and the same content, individual review probably isn't necessary.

/lgtm

/hold

Before this is merged, please install a cluster-bot cluster with this PR and with https://github.com/openshift/openshift-apiserver/pull/209

**Pull request closed without merging**